### PR TITLE
Fix Dart syntax highlighting issue

### LINKED ISSUE #8151: Improve Dart syntax highlighting
### Check for existing issues

- [X] Completed

### Describe the feature

Awesome to have Dart support in https://github.com/zed-industries/zed/pull/7220 !

I did notice however that the syntax highlighting is somewhat lacking compared to VSCode as seen in these screenshots:

VSCode

Screenshot 2024-02-21 at 21 48 39

Zed
Screenshot 2024-02-21 at 21 46 19

Would it be possible to improve this?

Hope you don't mind if i ping you in this issue @agent3bood as you made the initial Dart PR

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by Imgkl**
Release Notes:
* Improved Dart syntax highlighting ([#8151](https://github.com/zed-industries/zed/pull/8347#8151))

Before:
Before

After:
After

(Shameless plug, since I build the theme)
Theme: The Dark Side

### CODE REVIEW

**Comment by Imgkl** [comment]
Before the release of version 0.123.4, integrating this PR would be beneficial as it enhances syntax highlighting for Flutter and Dart applications.

**Comment by MarlonJD** [comment]
That's awesome, it's improving reading the code. LGTM. I'm using preview version, if it's won't merge soon, I'll cherry-pick and use this.

**Comment by Imgkl** [comment]
@MarlonJD, thanks a ton for checking out the PR! 🌟 Crossing my fingers it gets the green light soon!

As for the Flutter debugger, that’s not my jam right now. There’s an open issue where the pros are hashing it out. So, I’m stepping back to let those in the know do their thing. 😁

**Comment by Imgkl** [comment]
@maxbrunsfeld any timeline on when we can get this PR reviewed? Are we waiting for language support as extension?

**Comment by Imgkl** [comment]
I've started the work for transitioning dart support as extension. Currently the blocker is the lack of language servers api.

File: crates/languages/src/dart/locals.scm [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+[
+  (block)
+  (if_statement)
+  (for_statement)
+  (while_statement)
+  (try_statement)
+  (catch_clause)
+  (finally_clause)
+] @local.scope
```
Comment (maxbrunsfeld) [inline crates/languages/src/dart/locals.scm:32]: This isn't a file that Zed uses. Can you remove this?

File: crates/languages/src/dart/indents.scm [hunk 2]
```diff
@@ -1,7 +1,45 @@
...
+
+(return_statement
+  ";" @indent.end)
+
+(break_statement
+  ";" @indent.end)
+
+(comment) @indent.ignore
+
+(if_statement) @indent.auto
```
Comment (maxbrunsfeld) [inline crates/languages/src/dart/indents.scm:45]: This looks like a file that's been copied from elsewhere. maybe nevoid? Zed doesn't use this same indent syntax. Can you undo the changes to this file?

File: crates/languages/src/dart/folds.scm [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  (enum_declaration)
+  (extension_declaration)
+  (arguments)
+  (function_body)
+  (block)
+  (switch_block)
+  (list_literal)
+  (set_or_map_literal)
+  (string_literal)
+] @fold
```
Comment (maxbrunsfeld) [inline crates/languages/src/dart/folds.scm:12]: Zed currently doesn't use a query called `folds.scm`. Can you remove this?

**Review by maxbrunsfeld — CHANGES_REQUESTED** [review]

File: crates/languages/src/dart/folds.scm (hunk 3 above)
Comment (Imgkl) [reply crates/languages/src/dart/folds.scm]: Oh. I saw folds schema added to languages like Zig, Erlang and Svelte.


image


I've removed it as you suggested, but let me know, if we can have it again.

File: crates/languages/src/dart/locals.scm (hunk 1 above)
Comment (Imgkl) [reply crates/languages/src/dart/locals.scm]: Understood. Removed.

File: crates/languages/src/dart/indents.scm (hunk 2 above)
Comment (Imgkl) [reply crates/languages/src/dart/indents.scm]: I kept neovim as base and made few minor changes. I've reverted the file contents now.

**Comment by Imgkl** [comment]
@maxbrunsfeld just checking in, any updates on this?

**Comment by maxdeviant** [comment]
@Imgkl We've extracted the Dart support out into an extension: 

https://github.com/zed-industries/zed/tree/3a6e0bb9b68ddb8c471c71b25d3052f6fae8b703/extensions/dart

Could you rebase this PR against `main` to account for the change in file locations?

**Comment by Imgkl** [comment]
Sure, will do. @maxdeviant

**Comment by Imgkl** [comment]
@maxdeviant Done. This branch is rebased with main and pushed my changes.

**Review by maxbrunsfeld — APPROVED** [review]

**Comment by JoseDFS** [comment]
what version of zed has this?

**Comment by maxdeviant** [comment]
> what version of zed has this?

This hasn't been released yet, but will be in v0.0.2 of the Dart extension: https://github.com/zed-industries/zed/pull/10940

I'll get that published soon.
